\documentclass[titlepage,11pt]{article}
\usepackage[dvips]{graphicx}    % needed for including graphics e.g. EPS, PS
\usepackage{epsfig}
\usepackage{url}
\usepackage{helvet}
%\usepackage{natbib}
\usepackage[comma,sort]{natbib}
%\usepackage{lscape}
%\usepackage[dvips=true,bookmarks=true]{hyperref}
%\topmargin -1.5cm        % read Lamport p.163
%\topmargin .25cm        % tex automatically gives 1 inch, we need 1.86cm
%\topmargin 1.86cm        % tex automatically gives 1 inch, we need 1.86cm more to achieve utopia
%\margin 4.4cm        % read Lamport p.163
%\oddsidemargin 2.5cm   % read Lamport p.163
%\evensidemargin 2.5cm  % same as oddsidemargin but for left-hand pages
%\oddsidemargin 2.5cm
%\evensidemargin 2.5cm

%\textwidth 11.5cm  % official CWCE format
%\textwidth 13.5cm
%\textheight 19cm   % official CWCE format
%\pagestyle{empty}       % Uncomment if don't want page numbers
%\parskip 5pt           % sets spacing between paragraphs
% 1.5 spacing gives 27 lines per page in helvetica and times 11pt font
\renewcommand{\baselinestretch}{1.5} % Uncomment for 1.5 spacing between lines
%\parindent 0pt  % sets leading space for paragraphs

% adobe times-roman
%\newfont{\titfont}{ptmr at 16truept}
%\newfont{\autfont}{ptmro at 11truept}
%\newfont{\absfont}{ptmb at 12truept}
%\newfont{\bodfont}{ptmr at 10truept}

% some other times
%\newfont{\titfont}{ctmr at 16truept}
%\newfont{\autfont}{ctmri at 11truept}
%\newfont{\absfont}{ctmb at 12truept}
%\newfont{\bodfont}{ctmr at 10truept}

% Y.A. times
\newfont{\titfont}{tir at 16truept}
\newfont{\autfont}{tir at 12truept}
\newfont{\afffont}{tio at 11truept}
\newfont{\absfont}{tib at 12truept}
\newfont{\bodfont}{tir at 10truept}
\newfont{\bibfont}{tir at 8truept}
\newfont{\bibitfont}{tio at 8truept}

\begin{document}
\fontfamily{phv}\selectfont

% Start your text
\title{Mauve: Multiple Alignment of Conserved Genomic Sequence}

\author{
Aaron C. E. Darling
\footnote{1656 Linden Dr., Madison, WI 53706 USA, darling@cs.wisc.edu}
\thanks{Department of Computer Science}
\thanks{Department of Animal Health and Biomedical Sciences}
\and
Bob Mau$^3$
\thanks{Department of Oncology}
\and
Frederick R. Blattner
\thanks{Department of Genetics}
\thanks{Genome Center of Wisconsin, University of Wisconsin-Madison, USA}
\and
Nicole T. Perna$^3$ $^6$\\
Running Title -- Multiple Alignment of Conserved Genomic Sequence\\
Keywords:  genome alignment, comparative genomics, breakpoint elimination,\\
genome rearrangements
}
\date{ \today }

%\footnote{}

\maketitle

%\begin{abstract}
%\absfont

\begin{abstract}
As genomes evolve, they undergo large scale evolutionary processes that present
a challenge to sequence comparison not posed by short sequences. Recombination
causes frequent genome rearrangements,
horizontal transfer introduces new sequences into bacterial chromosomes, and
deletions remove segments of the genome. Consequently, each genome is a mosaic
of unique lineage-specific segments, regions shared with a subset of other
genomes and segments conserved among all the genomes under consideration.
Further, the linear order of these segments may be shuffled among genomes.

We present methods for identification and alignment of conserved genomic DNA in the
presence of rearrangements and horizontal transfer. Our methods have been
implemented in a software package called Mauve.  Mauve has been applied to align
9 enterobacterial genomes and to determine global rearrangement structure in 3
mammalian genomes.  We have evaluated the quality of Mauve alignments and drawn
comparison to other methods through extensive simulations of genome evolution.

\subsection*{Availability}
Source code and binaries are freely available for academic and non-profit
research. Commercial licenses are also available.  See
\url{http://gel.ahabs.wisc.edu/mauve} for more details.
\end{abstract}


\section*{ Introduction }
%Introduction - Why sequence alignment exists - original problem
%  - comparing individual gene sequences
%  - identify residue changes
%  - identify small insertions and deletions
%- introduction of genome sequencing
%  - methods developed to deal with genomic scale sequences
%  - aligners for long sequences (MUMmer, etc.)
%  - aligners for multiple long sequences (MGA, MLAGAN, MultiPipMaker)
%- determining pairwise genome structure: Shuffle-LAGAN - current

The recent determination of numerous bacterial and eukaryotic genome sequences
poses new challenges for comparative sequence analysis.  In addition to
identifying local changes in the sequences of individual genes, the
availability of genome sequences provides a basis for comparison of the
structure and organization of genomes as a whole.  Genomes are known to undergo
several types of large-scale evolutionary events.  Gene duplication can result
in the existence of paralogous genes, while gene loss may remove a copy and
obscure the assumption of orthology.  Reordering of genetic elements occurs by
mechanisms such as repeated inversion or translocation.  Horizontal transfer
introduces new genetic elements into bacterial genomes~\citep{hacker_ht}.  Consequently, each
genome is a mosaic of unique lineage-specific segments, regions shared with a
subset of other genomes and backbone segments conserved among all the genomes
under consideration.  Further, the linear order of these segments will be
shuffled among genomes~\citep{bourquepevzner,sankoff}. Genome comparison
systems must account for all of these evolutionary phenomena in order to
provide a complete picture of genetic differences among
organisms.



Early sequence comparison methods were designed to identify nucleotide
substitutions and small insertions and deletions by computing an alignment of
pairs of short sequences.  Such early techniques as Needleman-Wunsch global
alignment and Smith-Waterman local alignment employ methods whose computation
time scales as $O(n^2)$ where $n$ is the length of input sequences.  Numerous
multiple sequence alignment
and comparison methods are based on dynamic programming algorithms similar to
Smith-Waterman and
Needleman-Wunsch~\citep{clustalw,poalign,tcoffee,dialign2,Mor96PNASUSA}.  
Such pairwise and multiple sequence alignment methods suffer the limitation
that application to long (typically $n
>$ 10Kbp) sequences is prohibitively time-consuming.

The availability of genome sequences demands methods for aligning long genomic
DNA sequences.  Several heuristic approaches to align long sequences have been
developed under the assumption that highly similar subsequences can be found
quickly and are likely to be part of the correct global alignment.  These local
alignments are used to anchor a global alignment, reducing the number of
possible global alignments considered during a subsequent $O(n^2)$ dynamic
programming step.  Some spurious local alignments are typically found due to
random sequence similarity, particularly when using a sensitive local alignment
method.  A method for selecting alignment anchors must be
employed to filter out spurious matching regions. Alignment tools such as
MUMmer, GLASS, AVID, and WABA align pairs of long sequences,
implementing various methods to discover local
alignments~\citep{avid,glass,mummer,waba,segmentgenomic}. Similar
\textit{multiple} sequence alignment methods for long sequences have been developed and
implemented in software packages such as MAVID,
MLAGAN, and MGA~\citep{mga,mavid,lagan}.  All of these pairwise and multiple sequence
aligners assume the input sequences
are free from significant rearrangements of sequence elements,
selecting a single collinear set of alignment anchors.

Recently, methods have been developed to perform pairwise genome comparison in
the presence of rearrangements.  
Shuffle-LAGAN, a variant of the LAGAN alignment
system, was the first genome comparison
method described that explicitly deals with genome rearrangements during the
alignment process~\citep{slagan}.  Like other genome alignment methods, Shuffle-LAGAN uses an
anchored alignment approach.  Rather than selecting a single collinear set of
anchors, Shuffle-LAGAN selects anchors collinear in the first sequence
with rearrangements permitted in the other sequence.  Its selection criteria
optimizes a scoring metric that includes penalties for specific classes of
rearrangements such as translocations, inversions, and translocated inversions
in addition to the traditional gap and substitution penalties.  The resulting
set of anchors, referred to as a 1-monotonic map, forms the basis for further
alignment.  Although Shuffle-LAGAN's alignment approach appears to work well for
pairwise comparison, an extension of the method to multiple genome sequences has
not yet been suggested.

MultiPipMaker, based on \textit{blastz}, is a tool that can align multiple genomes
to a single reference genome in the presence of rearrangements~\citep{multipipmaker}.  
MultiPipMaker uses \textit{blastz}~\citep{blastz} on each pair of reference and
non-reference genomes to calculate pairwise local alignments.  These local
alignments are used to construct a rough global alignment that is iteratively
refined.  Because MultiPipMaker does not provide a mechanism for global alignment of
regions not included in the initial local alignments, more divergent homologous
regions between local alignments may remain unaligned.  As such, MultiPipMaker
can best be described as a multiple local aligner for genome sequences, rather
than a global aligner. 
Furthermore, neither
Shuffle-LAGAN nor MultiPipMaker provide a means to precisely identify the
breakpoints of multiple genome rearrangements.

%problem: 9 enterobacterial genomes [reference figure with genomes]
%  - comparing structure of entire genomes
%  - identify, in addition to the above:
%    - conserved orthologous regions (backbone)
%      - cite perna
%    - paralogous regions
%    - horizontally transferred sequence regions
%    - changes in gene order: inversions and translocations
%      - cite parkhill, pestis paper
%      - identifying such collinear regions gives rise to a permutation matrix
%      that can be used for inversion distance and evolutionary reconstruction
%    - recombinational breakpoints

During the past several years, groups from around the world published the
finished genome sequences of 9 enterobacteria
(Table~\ref{table:9_entero}).  Previous studies have shown that these 9
enterobacterial genomes have undergone significant horizontal transfer and
numerous genome rearrangements since their divergence.  However,
a lack of effective tools has constrained comparison of the rates and patterns
of large scale evolutionary processes in these bacteria to pairwise and 3-way
studies.

%- Mauve
%  - We describe a genome comparison method that identifies regions of genomic
%  backbone, rearrangements and inversions in backbone regions, the exact
%  sequence breakpoints of such backbone rearrangements.  Furthermore, our
%  comparison method performs traditional multiple alignment of conserved regions
%  to identify nucleotide substitutions and indels.  We implemented our methods
%  in a genome alignment package called Mauve.
%  - genome rearrangements that meet a minimum weight criteria (described in
%  Methods section) are identified
%  - rearranged regions must be present in all sequences under consideration
%  - first sequence is used to assign a reference orientation for sequence
%  regions
%  - we developed a viewer to display the pattern of genome rearrangements

We describe a genome comparison method that identifies conserved genomic
regions, rearrangements and inversions in conserved regions, and the exact
sequence breakpoints of such rearrangements across multiple genomes.
Furthermore, our comparison method performs traditional multiple alignment of
conserved regions to identify nucleotide substitutions and small insertions and
deletions (indels).  We implemented our methods in a genome alignment package
called Mauve.  Mauve represents the first alignment system that integrates
analysis of large-scale evolutionary events with traditional multiple sequence
alignment.  By integrating these previously separate analysis steps, Mauve
provides additional ease-of-use and sensitivity over other systems when
comparing genomes with significant rearrangements.

Like other genome alignment methods, Mauve uses anchoring as a heuristic to
speed alignment.  Unlike other multiple genome alignment systems, Mauve's
anchor selection method relaxes the assumption that the genomes under study are
collinear.  Instead, Mauve identifies and aligns regions of local collinearity
called locally collinear blocks (LCBs).  Each locally collinear block is a
homologous region of sequence shared by the genomes under study.  An LCB may
contain regions unique to a subset of the genomes but does not
contain any rearrangements of homologous sequence shared by all genomes.

The locally collinear blocks identified by Mauve's anchor selection algorithm
are required to meet a user specified minimum weight criteria as described in
the Methods section.  The weight of an LCB provides a measure of confidence
that it is a true genome rearrangement rather than a spurious match.
By selecting a high minimum weight during alignment the user can identify
genome rearrangements that are very likely to exist, whereas by selecting a
lower minimum weight the user can trade some specificity for sensitivity to
smaller genome rearrangements.

In addition to the alignment algorithm, a simple viewing system has been
developed to display the rearrangement structure of several genome sequences.
The viewer uses the first sequence to assign a reference orientation to LCBs in
the remaining sequences.  Thus regions that are in the reverse-complement
orientation relative to the first sequence appear inverted in the viewer.
Because the boundaries of rearrangement have been determined, the viewer is
able to draw a single line that logically connects the entire homologous
collinear blocks
from each genome.  Previous visualization systems drew one line per local
alignment often yielding a confusing picture of complex rearrangement
structures.

%- Evaluating the quality of genome alignments
%  - no genome-scale benchmark data sets exist to assess the quality
%    of alignments with rearrangements
%  - we developed a simple genome evolution simulator that evolves the correct
%    alignment
%  - the quality of calculated alignments can be scored using the correct
%  alignment of the simulated data.

Finally, to make an informed decision when choosing between alignment tools, it
is important to have not only an understanding of the algorithms used but also
the empirical performance of the alignment system.  Towards this end we
empirically characterized our alignment system and compared its performance
with other well known genome alignment systems.  
Manually validating a benchmark alignment on the genome scale is too labor
intensive.  Instead we developed a simple
genome evolution simulation system that incorporates large and small scale
evolutionary events.  Because the evolutionary history is known, the simulator
can generate the 'correct' alignment in addition to the evolved sequences.  We
measured the ability of Mauve and other genome aligners to reproduce the
'correct' alignment for the evolved sequences.



\section*{ METHODS }


%- design considerations
%  - must handle long genome sequences
%    - use anchored alignment approach
%  - need to avoid spurious anchors in paralagous regions:  use MUMs
%  - miss many anchors due to repetivity or polymorphisms
%  - use recursive anchoring to get additional anchors not initially found
%  - must handle reordering of homologous regions:
%    - traditional maximum weight consistent set approaches don't work
%    - select anchors using a method that handles reordering
%   - use a greedy algorithm to approximate a maximum weight set of anchors
%     such that each consistent subset of anchors meets a minimum weight
%     requirement
%  - a full gapped alignment of intervening regions should be performed
%    - use ClustalW because its behavior is well understood
%  - because progressive alignment is used, a guide tree is needed.
%    rather than recalculating the guide tree during each dynamic programming
%   step, infer a guide tree for the genome sequences

The set of target genomes for our alignment system led us to consider several
factors when designing an alignment algorithm.  The alignment system must
quickly align long genome sequences.  Although parallel dynamic programming
methods have been used with some success~\citep{paralleldp}, anchored
alignment approaches require only modest computational resources while having
a minimal impact on alignment quality.

The target genomes are known to have significant repetitive regions such as
ribosomal RNA operons and prophages.  When searching for anchors across
multiple genomes, problems arise if a particular repetitive motif occurs
numerous times in each sequence because it becomes unclear which combination of
regions to align.  For a repetitive element existing $r$ times in each of $G$
genomes, there will be $r^G$ possible alignment anchors, of which at most $r$
represent truly orthologous anchors.  Mauve avoids this
problem by using Multiple Maximal Unique Matches (multi-MUMs) of some minimum
length $k$ as alignment
anchors.  Multi-MUMs are exactly matching subsequences shared by two or more
genomes that occur only once in those genomes and that are bounded on either
side by mismatched nucleotides.  Because using multi-MUMs reduces
anchoring sensitivity in conserved repetitive regions and regions that have undergone
numerous nucleotide substitutions or indels, Mauve employs a recursive anchoring
strategy that progressively reduces $k$, searching for smaller anchors between
existing alignment anchors.

The enterobacterial genomes are known to have undergone significant genome
rearrangements as described in their genome papers.  Algorithms used by other
global multiple
alignment systems anchor their alignments by selecting the highest scoring
collinear chain of local alignments~\citep{mga,mavid}.  Such methods preclude 
identification of the rearrangements known to exist in our data
set and many others. To
successfully align our target genomes, the anchor selection method should
identify consistent (collinear) subsets of local alignments to use as anchors
while filtering out unlikely local alignments. Ideally, an algorithm would
identify a maximum-weight set of anchors such that each collinear subset of
anchors meets some minimum-weight criteria.  Mauve uses a greedy breakpoint
elimination algorithm to
generate an approximate solution to the maximum-weight non-collinear anchoring
problem.

To align the intervening regions of sequence between anchors our method employs
the progressive dynamic programming approach of Clustal-W~\citep{clustalw}.  In progressive
alignment, a phylogenetic guide tree specifies the optimal progression of
sequences to align when building the multiple alignment.  Rather than
recalculating a guide tree during each alignment of intervening regions, Mauve
infers a single global phylogenetic tree.  

The alignment algorithm can be summarized as follows:
\begin{enumerate}
\item Find local alignments (multi-MUMs)
\item Use the multi-MUMs to calculate a phylogenetic guide tree
\item Select a subset of the multi-MUMs to use as anchors
\item Partition the anchors into collinear groups called LCBs.
\item Perform recursive anchoring to identify additional alignment anchors
\item Perform a progressive alignment of each LCB using the guide tree
\end{enumerate}

\subsection*{Finding multi-MUMs}

%- Finding Putative Anchors
%  - Mauve uses a simple anchoring method based on Maximal Unique Matches across
%    multiple sequences (multi-MUMs) as anchors in the anchored alignment process.
%  - Chose this method because it was sensitive enough for the target sequences
%  - The algorithm to find multi-MUMs is an extension of the seed-and-extend
%  hashing method used in GRIL and described in supplemental material by [GRIL ref].
%  - Basic overview:
%   - O(g*n^2), very efficient in practice %% could be O(g n log n)
%   - Construct sorted k-mer lists for each sequence for the specified k-mer size
%   - Scan SMLs looking for k-mers that occur at most once in each sequence
%   - If the diagonal represented by the k-mer match has not yet been discovered
%   then extend the diagonal until a mismatch occurs.
%   - when a mismatch occurs, try extending the diagonal in the subset of
%   sequences that still match, but only if the diagonal represented by the subset
%   match has not yet been discovered.
%   - A hash table is used to track known diagonals

The algorithm to find multi-MUMs is an extension of the simple seed-and-extend
hashing method used by \texttt{GRIL}~\citep{gril}.  While the seed-and-extend
algorithm has time complexity $O(G n^2)$ where $G$ is
again the number of genomes and $n$ the length of the longest genome, it is very
fast in practice.  Further, the random-access memory requirements are proportional to the
number of multi-MUMs found, not $n$, allowing it to efficiently tackle large
data sets.  $O(G n)$ disk space is used to store sequentially accessed data
structures.

Briefly, the algorithm proceeds by constructing a sorted list of $k$-mers for
each genome $g \in G$.  The sorted $k$-mer lists are then scanned to identify
$k$-mers that occur in two or more sequences but that occur at most once in any
sequence.  If a multi-MUM that subsumes the $k$-mer match has not yet been
discovered, then the match
seeds an extension in each genome until a mismatch occurs.  When a mismatch
occurs an extension is seeded in the subset of sequences that are still
identical, but only if a subsuming multi-MUM has not yet been discovered.

Formally we define each multi-MUM as a tuple $\langle L, S_1,...S_G\rangle$ where
$L$ is the length of the multi-MUM, and $S_j$ is the left-end position of the
multi-MUM in the $j^{th}$ genome sequence.
We denote the resulting set of multi-MUMs as $\mathbf{M} = \{M_1...M_N\}$.  The
$i^{th}$ multi-MUM in $\mathbf{M}$ is referred to as $M_i$.  To refer to the length of
$M_i$ we use the notation $M_i.L$ and similarly, we refer to the left end of
$M_i$ in the $j^{th}$ input sequence using the notation $M_i.S_j$.  If
multi-MUM $M_i$ includes a region in the reverse complement orientation in sequence
$j$, we define the sign of $M_i.S_j$ to be negative.  Finally, if multi-MUM
$M_i$ does not exist in sequence $j$, we define $M_i.S_j$ to be 0 -- the
left-most position in any genome is 1 (or -1).

%- Calculating a guide tree
%  - Use the ratio of base pairs shared between two sequences to their average
%  sequence length as a distance metric.
%  - construct a tree using neighbor-joining [ref!]
%  - The MUMs found in the previous step can be used to estimate the # or shared
%    base pairs, each matching base pair should be counted only once.
%  - Since MUMs can overlap each other, overlaps must be eliminated.
%  - Mauve resolves overlaps in favor of the higher multiplicity
%  match, where multiplicity is defined as the number of sequences the match
%  is defined in.
%  - When the multiplicity is equal, overlaps are resolved in favor of the longer
%  match.

\subsection*{Calculating a guide tree}
The method described to find multi-MUMs differs from that used by \texttt{GRIL} in that
it can identify multi-MUMs in subsets of the genomes under study.  Mauve
exploits the information provided by subset multi-MUMs as a distance metric to
construct a phylogenetic guide tree using Neighbor Joining~\citep{njoining}.
Specifically, the ratio of base pairs shared between two genomes to their average
genome length provides an estimate of sequence similarity.  This similarity
estimate is converted to a distance value for the Neighbor Joining distance
matrix by subtracting it from one.
Because multi-MUMs can overlap each other, calculating the similarity metric
requires that overlaps among multi-MUMs are resolved such that each matching
residue counts only once.  Mauve resolves overlaps in favor of the higher
multiplicity match, where $multiplicity(M_i)$ is defined as the number of genomes for
which $M_i.S_j \neq 0$.  If the multiplicity of two overlapping matches is
identical, the overlap is resolved in favor of the longer match.

%- Selecting a set of anchors
%  - This step removes spurious matches due to random sequence similarity and
%    polymorphic repetitive subsequences
%  - Mauve builds on the methodology developed in GRIL for determining the
%  boundaries of collinear regions of sequence called Locally Collinear Blocks.
%  - definition of LCBs
%  - In terms of matches, an LCB can be considered a consistent subset of
%  diagonals, e.g. a set of diagonals such that there exists a path through the
%  dynamic programming alignment matrix that includes all diagonals.
%  - Matches found in the previous step can be represented <l, s, .., s_G>
%  - formally, a consistent set of matches satisfies the total ordering property
%    blah blah

\subsection*{Selecting a set of anchors}
In addition to local alignments that are part of truly homologous regions, the
set of multi-MUMs $\mathbf{M}$ may contain spurious matches arising due to
random sequence similarity.  This step attempts to filter out such spurious
matches while determining the boundaries of locally collinear blocks.
An LCB can be considered a consistent subset of the multi-MUMs in $\mathbf{M}$.
Formally, an LCB is a sequence of multi-MUMs $lcb \subseteq \mathbf{M}$, $lcb =
\{M_1, M_2, ... , M_{|lcb|}\}$ that satisfies a total ordering
property such that $M_i.S_j \leq M_{i+1}.S_j$ holds for all $i$, $1 \leq
i \leq |lcb|$ and all $j$, $1 \leq j \leq G$.  For a given set of multi-MUMs,
the minimum partitioning of $\mathbf{M}$ into collinear blocks can be found
through breakpoint analysis~\citep{breakpointPhylogenies}.  Breakpoint analysis
requires that matching regions exist in \textit{all} genomes under study, so
multi-MUMs with multiplicity less than $G$ are removed from $\mathbf{M}$ before
performing this step of the algorithm.

%  - Mauve uses a greedy algorithm to remove low-weight consistent subsets.
%  First Mauve performs the following 3 steps repeatedly:
%   1. determine consistent subsets of M
%   2. calculate the weight of each consistent subset w(cs_i)
%   3. identify the collinear subset(s) MCS that satisfy min( W(cs_i ) )
%   3. Stop if min( W(cs_i) ) >= MinimumWeight
%   4. remove all matches m member MCS from M
%  - Here W(cs) is defined as the sum of weights of each match m member cs.
%    and w(m) is defined as the product of the multiplicity and the length of the
%   match
%  - The resulting collinear sets of anchors define the Locally Collinear Blocks that
%    are used to guide the remainder of the alignment process.

Given a minimum weight criteria $MinimumWeight \geq 0$, Mauve uses a greedy
breakpoint elimination algorithm
to remove low-weight collinear subsets of $\mathbf{M}$.  Mauve performs the
following steps repeatedly until all collinear subsets of $\mathbf{M}$ meet the
minimum weight requirement:
\begin{enumerate}
\item Determine a partitioning of $\mathbf{M}$ into collinear subsets
$\mathbf{CS}$
\item Calculate the weight, $w(cs_i)$ of each collinear subset $cs_i \in \mathbf{CS}$
\item Let $z = \min_{cs \in \mathbf{CS}}{ w( cs ) }$
\item Stop if $z \geq MinimumWeight$
\item Identify the collinear subsets $\mathbf{MinCS} \subseteq \mathbf{CS}$ that satisfy
$w( cs_i ) = z$.
\item For each $cs \in \mathbf{MinCS}$, remove each multi-MUM $M \in cs$ from $\mathbf{M}$
\item Go to step 1.
\end{enumerate}
Here $w(cs)$ is defined as $\Sigma_{M_i \in cs}{M_i.L }$. %\cdot multiplicity(M_i)
By default the $MinimumWeight$ parameter is set to $3k$, where $k$ is the
seed length used during the initial search for multi-MUMs.  We chose $3k$ as a
default minimum weight because it appears to filter the majority of spurious
matches in data sets we have evaluated.
Figure~\ref{fig:coalescent} illustrates the process of identifying collinear
sets of multi-MUMs and how removing a low-weight collinear
region can eliminate a breakpoint.
The resulting collinear sets of anchors delineate the LCBs that
are used to guide the remainder of the alignment process.


%- Extending the LCBs
%  - The anchor selection method may not be sensitive enough to detect the full
%    region of homology surrounding the LCBs.
%  - typically due to:
%    - repetitive regions preventing MUM detection
%    - more frequent nucleotide substitutions in the region
%  - try to find more anchors in regions between LCBs
%    - let previous_weight = sum( W( cs_i ) )
%    - concatenate intervening sequences
%   - locate MUMs in concatenated sequences
%   - translate MUMs back into their original sequence coordinates
%   - add new MUMs to M
%   - perform iterative removal of low-weight consistent subsets as above
%    - repeat this process until sum( W(cs_i) ) == previous_weight

\subsection*{Recursive anchoring and gapped alignment}
The initial anchoring step may not be sensitive enough to detect the full
region of homology within and surrounding the LCBs.  
In particular, repetitive regions and regions with frequent nucleotide
substitutions are likely to lack sufficient anchors for complete alignment.
Using the existing anchors as a guide, two types of recursive anchoring are
performed repeatedly.  First, regions outside of LCBs are searched to extend the
boundaries of existing LCBs and identify new LCBs.  Second, unanchored regions
within LCBs are searched for additional alignment anchors.

When searching for additional anchors outside existing LCB boundaries, two
factors contribute to Mauve finding additional anchors.  First, Mauve uses a
smaller value of the match seed size $k$.  Second, only the regions outside
existing LCB boundaries are searched, so regions not unique in the
entire genome may be unique within regions outside LCBs.
Not only can the range of existing LCBs be extended by searching regions outside
LCB boundaries, but new LCBs that meet the minimum weight requirement can
be identified as well.  
To perform the search, the
outside sequences in each genome are concatenated into a single sequence per
genome.  We refer to the set of concatenated sequences as $\mathbf{S}$ and the
concatenated sequence from the $j^{th}$ genome as $S_j$.  Multi-MUMs of minimum length $k$ are found, where $k =
seed\_size(\mathbf{S}) - 2$, and 
$seed\_size(\mathbf{S}) = \log_2{(\Sigma_{j=1}^{G}{\frac{length(S_j)}{G}})}$. 
Because the left-end coordinates of each new
multi-MUM are defined in terms of the concatenated sequence they must be
transposed back into the original coordinate system.  The transposed
multi-MUMs are added to $\mathbf{M}$
and iterative removal of low-weight collinear subsets is performed as above.
The process of searching regions outside LCBs is repeated until $\Sigma_{cs \in
\mathbf{CS}}{w(cs)}$ remains the same during two successive iterations of the search.


%- Performing a gapped alignment
%  - Initial set of alignment anchors may leave large regions unanchored
%  - perform recursive anchoring on these regions
%  - select k based on log_2 ( max( length of intervening region ) )
%  - select the highest weight LCB in the intervening region
%  - search for anchors between the new matches
%  - stop anchoring when the intervening region is less than 200? bp
%  - send remaining region to ClustalW for progressive alignment using the
%    genome guide tree

In addition to missing anchors outside the boundaries of LCBs, the initial
anchoring pass may have lacked the sensitivity to find anchors in large regions
within each LCB.  Because progressive alignment requires relatively dense
anchors (at least one anchor per 10Kbp of sequence),  Mauve performs recursive
anchoring on the intervening regions between each pair of existing anchors.  Not only 
does this step anchor more divergent regions of sequence, it also locates anchors in
conserved repeats because many $k$-mers that are not unique in the whole genome are
likely to be unique within the intervening regions between existing anchors.
  Unlike
other genome aligners which perform a fixed number of recursive passes with a
pre-determined sequence of anchor sizes, Mauve calculates a minimum anchor size
based on the length of the intervening sequence and stops recursive anchoring
when either no additional anchors are found or when the intervening region is
shorter than a fixed length, defaulting to 200bp.  During each recursive anchor
search, a single collinear set of new anchors in the same orientation as the
flanking anchors is selected to cover the region
between flanking anchors. For each search, $k$ is calculated as above: $k =
seed\_size(\mathbf{S})$ where $\mathbf{S}$ is
the set of intervening sequences, one per genome.  By dynamically
calculating the value of $k$, Mauve ensures that $k$ is sized appropriately for
the intervening region.  Selecting $k$ too large prevents discovery of
multi-MUMs in polymorphic regions, whereas selecting $k$ too small increases
the likelihood that $k$-mers will not be unique in the intervening region.

Armed with a complete set of alignment anchors, Mauve performs a Clustal-W
progressive alignment using the genome guide tree calculated previously.  The
progressive alignment algorithm is executed once for each pair of adjacent
anchors in every LCB, calculating a global alignment over each LCB.  Tandem
repeats less than 10Kbp in total length are aligned during this phase.  Regions
larger than 10Kbp without an anchor are ignored.

\section*{ RESULTS }
The Mauve genome alignment procedure results in a global alignment of each
locally collinear block that is conserved among all the sequences under study.
The alignment identifies homologous regions conserved among all genomes
aligned.  Nucleotides in any given genome are aligned to at most one nucleotide
in each of the other genomes, suggesting orthology among aligned residues. 
The remaining unaligned regions may be lineage-specific sequence, or orthologous
or paralagous repetitive regions and can be identified as such during subsequent
processing with other tools.  Large ( $>$ 10Kbp) regions introduced to a subset
of the genomes by horizontal transfer are not aligned by Mauve because
they do not have alignment anchors conserved among all sequences.  Both large
and small regions existing in only a subset of the genomes and that also underwent
local rearrangement remain unaligned.
% - Result will be a global alignment of each locally collinear block
% - each base pair is aligned to at most one other residue per sequence
% - generates alignments of orthologous regions conserved

% - Several experiments were performed to evaluate alignment quality and
%   performance on real data.  Two of these experiments were designed to compare
%   Mauve to two other alignment systems, Multi-LAGAN and Shuffle-LAGAN
%   Multi-LAGAN was designed for cross-species genome comparison.  Its sensitive
%   match seeding based on the CHAOS local alignment algorithm was intended to
%   enable Multi-LAGAN to identify anchors in the presence of significant
%   nucleotide substitutions and indels.  By comparison, Mauve uses a very
%   insensitive match seeding technique.  We designed this experiment 
%

\subsection*{Evaluating alignment quality}
Without a 'correct' alignment of the 9 enterobacterial genomes, the calculated
alignment generated by Mauve can not be evaluated for accuracy.  In fact, no
manually curated multiple alignment benchmark data sets account for genome scale
evolutionary events such as inversion, rearrangement, and horizontal transfer.
Despite the lack of a manually curated correct alignment, we can estimate the
alignment accuracy by modeling evolution and aligning the simulated data set.

The inferential power yielded by evaluating alignment accuracy using simulated
evolution is only as strong as the degree to which the simulation faithfully
represents the actual
evolutionary processes that governed the history of the genomes under study.
Keeping that fact in mind, we constructed a simplistic model of genome
evolution that we believe captures the major types, patterns, and frequencies
of events in the history of the enterobacterial genomes.  Given a rooted
phylogenetic tree and an ancestral sequence we would like to generate evolved
sequences for each internal and leaf node of the tree, along with a multiple
sequence alignment of regions conserved throughout the simulated evolution. To
effectively represent genome evolution, the simulation must include nucleotide
substitutions and indels in addition to genome scale events such as horizontal
transfer, inversion, and rearrangement.

%- Evaluating quality of the alignments
%  - we would like to know how well Mauve aligned the 9 enterobacteria
%  - don't have the correct alignment, so simulate evolution to get an estimate
%  - The estimate of alignment quality is accurate only in-so-far as our model
%  accurately describes enterobacterial evolution
%  - Also compare Mauve to other alignment systems

Nucleotide substitutions are ostensibly the best understood and most ubiquitous
evolutionary mechanism. We use the HKY model of nucleotide substitution implemented
in the Monte-Carlo simulation package called Seq-gen~\citep{seqgen}.  Small
insertions and deletions (indels) are modeled as occurring with uniform
frequency and distribution throughout the genomes, with a size sampled from a
Poisson distribution with mean value 3bp. When studying the differences
between \textit{E. coli} O157:H7 EDL933 and K-12 MG1655, it became clear that a
small number of horizontal transfers introducing large regions of sequence have
occurred, while the majority of transfers introduced small sequence regions.
Our model includes large horizontal transfer events uniformly distributed in
length between 10Kbp and 60Kbp.  The size of small horizontal transfer events
is sampled from an exponential distribution with mean value 200bp.  Using the
observation that two overlapping inversion events can result in a genome
rearrangement, our model does not explicitly implement rearrangements. The
length of inversions are sampled from an exponential distribution with mean
value 50Kbp.  Locations for inversion and horizontal transfer events are
sampled uniformly throughout the genome, and all events are simulated to have taken
place at a point in time given by a marked Poisson process over the phylogenetic tree. 
Finally, genome size is expected to
stay relatively constant over time, so deletion events are sampled with the
same size and frequency as events that introduce new sequence.  Our
implementation of the evolutionary model described above is referred to as the
simple genome evolver, or just \texttt{sgEvolver}.  

%- simple genome evolver
%  - supports types of evolutionary events:
%    - nuc. substitutions ala seq-gen
%    - indels size~Poi(3bp)
%    - small horizontal transfer: exp size dist. 200bp
%    - large horizontal transfer: uniform size dist. 10Kbp-60Kbp
%    - inversions: exp size dist. 50Kbp
%  - model assumes genome size is constant on average so deletions occur with
%  same size/frequency as insertions (indels, small ht, and large ht)

\subsection*{Experiments}
Using the simple genome evolver, we designed and executed several experiments
to compare the ability of Mauve and other alignment systems to align
our target data set.  Multiple alignment experiments used the phylogenetic guide tree
estimated for the 9 enterobacteria, midpoint rooted to provide an entry point
for the ancestral sequence.  Rather than generate a random ancestral sequence,
1Mbp of enterobacterial DNA was used to preserve the distribution of sequence
motifs and repetitive subsequences found in our data set.  An additional 1Mbp of
enterobacterial DNA was used as a donor sequence pool for insertion events.

Three experiments were performed, each of which consists of numerous
simulations.  The first experiment evaluates the robustness of Mauve and
Multi-LAGAN, a cross-species genome comparison tool, to genomes with high
nucleotide substitution and indel rates.  A second experiment compares Mauve to
Shuffle-LAGAN when aligning pairs of genomes with rearrangements.  At the time
these experiments were performed, Shuffle-LAGAN was the only publicly available
genome aligner capable of aligning genomes in the presence of rearrangement. 
Our final experiment evaluates the ability of Mauve to align simulated genomes 
that resemble the 9 target enterobacteria.

For each simulated data set, alignments were calculated using the Condor high
throughput computing environment at the University of Wisconsin.  The
Wisconsin Condor cluster contains over 1000 nodes and allowed us to rapidly align
thousands of simulated data sets.
The calculated alignments were scored against correct alignments generated
during the evolution process.  We used the sum-of-pairs scoring procedure also
used by BaliBASE~\citep{balibase}.  In sum-of-pairs scoring,
each pair of aligned residues in the calculated alignment that are aligned to
each other in the correct alignment tallies a point.  The total alignment score
is then the ratio of points to total possible points.

%- Experiments
%  - evolution done using the phylogenetic guide tree estimated for the 9 enterobacteria
%  - midpoint rooted the tree
%  - each experiment evolved 1MB of enterobacterial DNA, using an additional 1MB
%  as a donor pool for insertions.
%  - calculated alignments scored against correct alignments
%  - used sum-of-pairs scoring procedure described in BaliBASE paper
%  - each pair of aligned residues in the calculated alignment that are aligned
%  to each other in the correct alignment tallies a point.  Total score is ratio
%  of points to possible points.
\subsubsection*{Mauve vs. Multi-LAGAN}
Our first experiment compared the ability of Mauve and Multi-LAGAN version 1.2
to align collinear sequences that had undergone increasing amounts of
nucleotide substitution and indels.  This experiment is designed to test the
sensitivity of the anchoring methods employed by each aligner.  We evolved 
9 genomes at 20 levels of nucleotide substitution and 20 levels
of indels, performing 2 replicate experiments of each combination of
substitution rate and indel rate.  The average Mauve and Multi-LAGAN alignment
accuracy for each simulation is displayed in
Figure~\ref{fig:ntsub_indel_mauve}.  From the figure, it is obvious that
Mauve's alignment score drops more rapidly than Multi-LAGAN's in the presence
of an increasing substitution rate.  We attribute this behavior to Mauve's use
of multi-MUMs as alignment anchors.  Multi-LAGAN's alignment anchors can
contain substitutions and indels, making them much more sensitive than exactly
matching subsequences.  At lower levels of nucleotide substitution, Mauve
appears to handle indels about as well as Multi-LAGAN.  For the nucleotide
substitution and indel rates previously reported in the enterobacterial data
set, Mauve aligns the simulated genomes with a high degree of accuracy.

%  - Nucleotide Substitution, Indels
%    - average of 2 runs
%    - Mauve vs. M-LAGAN
%    - MLAGAN is much more sensitive
%    - Mauve's use of MUMs limits its ability to align in the presence of subs.
%    - Mauve is more tolerant of indels than n. subs.
%    - if the sequences have no indels MLAGAN 1.2 crashes

\subsubsection*{Mauve vs. Shuffle-LAGAN}
We proceeded to gauge the ability of Mauve and Shuffle-LAGAN version 1.2 to
align sequences that had undergone increasing amounts of inversion and
nucleotide substitution. Because Shuffle-LAGAN is a pairwise aligner, we
reduced the number of taxa in our simulation from 9 to two.  Three simulations
were performed for each of 110 combinations of nucleotide substitution rate and
inversion rate.  The average accuracies of Mauve and Shuffle-LAGAN for each
experiment are shown in Figure~\ref{fig:ntsub_inv_mauve}.  Special
considerations must be taken when scoring Shuffle-LAGAN.  Because Shuffle-LAGAN
attempts to identify and align paralagous regions
a single residue in the first genome can be aligned to multiple
residues in the second genome.  For the purpose of scoring Shuffle-LAGAN, we
awarded points for a given residue in the first genome if any of the residues
in the second genome it was aligned to were correct.  

The experiment shows that
Mauve clearly excels at aligning rearranged sequences under lower substitution
rates that do not hamper its multi-MUM anchoring process.  Interestingly,
Shuffle-LAGAN appears to perform better as the substitution rate increases.
Based on our experience, we conjecture that this counter-intuitive result is related to
the repetitive nature of the ancestral enterobacterial sequence. Shuffle-LAGAN
appears to have difficulty selecting anchors in repetitive sequences.  As the
nucleotide substitution rate increases, regions that were repetitive are
randomly mutated and thus no longer repetitive.  Anchoring its alignment in
unique subsequences provides Mauve with immunity to this phenomena.

%  - NT sub vs. Inversions, pairwise
%    - average of 3 runs
%    - Mauve vs. Shuffle-LAGAN, only does pairwise, is the only other aligner
%    - if the sequences have no inversions Shuffle-LAGAN crashes
%    - special considerations for scoring Shuffle-LAGAN
%      - Shuffle-LAGAN may align bases of the first sequence to multiple residues
%        from the second sequence.  Credit was given if any of the aligned pairs
%        were correct.
%    - At low substitution rates Mauve clearly excels at aligning rearranged
%    sequences
%    - Shuffle-LAGAN appears to perform better as the substitution rate increses
%      - probably because the sequences are inherently somewhat repetitive and as nt sub
%      rate increases, the repetivity decreases.

\subsubsection*{An enterobacteria-like simulation}
Our final set of experiments sought to evaluate the ability of Mauve to align
genomes similar to the enterobacteria.  Evolutionary rates for the simulation
were extrapolated from previously published observations of the differences
between \textit{E. coli}
K-12 MG1655 and O157:H7 EDL933.  For these two \textit{E. coli}, there are about 75,000
observed nucleotide substitutions, about 4,000 observed indels,  40 large
horizontal transfer events, 400 small horizontal transfers, and one inversion.
The observed frequencies were converted to rates used to assign event
frequencies to branches of the phylogenetic guide tree.  It is known that among
the group of enterobacteria, the \textit{Salmonella} have higher rates of
inversion and rearrangement than the \textit{E. coli}.  To compensate, the
inversion rate was adjusted to result in approximately 30-40 inversion events.
When varying the substitution and indel rates between 0 and 125\% of the
observed rates while holding horizontal transfer and inversion rates constant,
Mauve alignments consistently average 80\% accurate, +/- 5\%.  The quality of
alignment does not appear to drop as the substitution and indel rates are
increased in this range.  Rather, it appears that horizontal transfer rates
have a more significant impact on alignment quality.  As horizontal transfer
rates increase, the ratio of lineage-specific sequence to backbone sequence
increases and Mauve's alignment algorithm aligns decreasing amounts of
the total sequence.  Figure~\ref{fig:entero_ht} shows how Mauve's ability to
align enterobacteria-like genomes changes as horizontal transfer rates
increase. When scored only against regions of the simulated genomes considered
as conserved backbone, Mauve consistently aligns with $>$98\% accuracy. For the
purpose of scoring the alignment, we define backbone as a region in the correct
alignment containing more than 50 gap-free columns without stretches of 50 or
more consecutive gaps in any single genome sequence. Based on our simulations
we believe our method accurately aligns the backbone of the 9 enterobacteria,
however, significant lineage-specific regions remain unaligned.

%  - enterobacteria-like
%    - average of 8 runs
%    - Mauve only since it's the only capable aligner
%    - estimated rates of each type of event based on the observed rates between
%      O157:H7 EDL933 and K12
%      - ntsub rate <blah>
%      - indel rate <blah>
%      - small ht rate <blah>
%      - large ht rate <blah>
%      - inversion rate <blah>
%    - Mauve is a backbone aligner, regions that are horizontally transferred do
%    not constitute backbone sequence
%    - The accuracy on the backbone region is very high, ~98%
%    - We can say with some confidence that Mauve aligns the backbone of the 9
%    enterobacteria accurately.

\subsection*{Alignment of 9 enterobacterial genomes}
We applied Mauve to align the 9 enterobacterial genomes listed in
Table~\ref{table:9_entero}.  Previous studies of these genomes indicates they
underwent significant genome rearrangement, horizontal transfer, and other
recombination~\citep{o157,styphi2}.  Mauve consumed 3 hours to align the 9 taxa
on a 2.4GHz computer with 1GB of RAM.  The alignment of the 9 taxa reveals
 45 LCBs with a minimum weight of 69.  
Figure~\ref{fig:entero_tree} shows the guide tree generated for these species.
The visualization of the genome rearrangement structure generated by the Mauve
viewer is shown in Figure~\ref{fig:entero_alignment}.  We can quickly visually
confirm several known inversions such as the O157:H7 EDL933 inversion relative
%to VT-2 sakai and K-12~\citep{pernachapter} 
to K-12~\citep{o157} 
and the large inversion about the
origin of replication among the
\textit{S. enterica} serovars Typhi CT18 and Ty2~\citep{styphi2}.

We proceeded to extract conserved backbone sequence from the alignment.  Again,
backbone is defined as regions of the alignment containing more than 50 gap-free
columns without stretches of 50 or more consecutive gaps in any single genome
sequence.  Under this definition, the 9 enterobacteria have 2.86Mbp of conserved
backbone sequence broken into 1252 backbone segments.  Across the backbone the
level of nucleotide identity is high, as shown by the identity matrix in
Table~\ref{table:entero_identity}.  
%\textit{would it be worth saying something
%here about the distribution of islands?  For example: Of the xx islands, yy
%appear to be sequence unique to a single taxon, suggesting recent acquisition by
%horizontal transfer.  zz of these islands exist in several taxa, suggesting
%differential gene loss as the as the mechanism underlying the observed
%difference.}
%- Mauve was applied to align 9 enterobacterial genomes that had undergone
%significant genome rearrangement, horizontal transfer, other recombination.
% - 9 enterobacteria genome characteristics
% - alignment properties
%  - x MB of conserved backbone sequence
%  - x number of LCBs
%  - x minutes on x computer, x MB of RAM

\subsection*{Rearrangements in 3 mammalian genomes}
Although we designed our methods with the intent of aligning bacterial genomes,
we applied Mauve to the entire mouse, rat, and human genomes to assess the
scalability of our methods.  For this experiment we used the ``finished''
human genome build 34, mouse genome build 32, and rat genome RGSC build
3.1.  Rather than complete a full alignment, Mauve was used to determine the
global rearrangement structure and LCBs in the three genomes.  Finding an
initial set
of anchors with minimum length 31bp consumed about 12 hours on a 1.6GHz
desktop workstation.  Computing the anchors consumes roughly 3GB RAM, however
the workstation was equipped with only 2.5GB of true memory, and disk-based virtual memory
was used to supply the remaining need.  Figure~\ref{fig:hmr} shows the
complex rearrangement structure of these three mammalian genomes. In this data
set it is difficult to determine the 'correct' number of LCBs: depending on the
minimum weight parameter used, the number of LCBs ranges from about 1000 to
2000.  Furthermore, the large minimum anchor size (31bp) precludes
identification of small, local rearrangements of the type previously reported by
\cite{slagan}. A full mouse-rat-human alignment using Mauve may help resolve the
true number of collinear blocks and facilitate identification of local
rearrangements, but has not yet been performed.

%- To assess scalability, Mauve was applied to human, mouse, and rat genomes
%  - finding an initial set of anchors takes 12 hours, about 3GB RAM
%  - number of LCBs ranges from 1k-2k depending on weight
%  - difficult to determine the 'correct' number of LCBs
%  - Full alignment seems possible but hasn't yet been performed

\section*{ DISCUSSION }

%- Seq alignment has progressed considerably
%- with advent of genome sequencing, the task definition is now different
%- Shuffle-LAGAN implements the first solution for automated alignment in the
%  presence of rearrangements
%- tools to automate the process of identifying rearrangements are
%  becoming more important now that high levels of rearrangement are being
%  identified in eukaryotic organisms. [cite pevzner]

Since their first application to molecular biology some 30 years ago, sequence
alignment techniques have progressed considerably.  With the advent of genome
sequencing a new type of sequence alignment problem, that of whole genome
comparison, has emerged.  Early approaches to genome alignment were designed to
tackle dramatically increased sequence lengths, but did not consider the
additional types of evolutionary events observed on the genome scale.  Genome
rearrangements, horizontal transfer, and duplication obfuscate
orthology.  As genomes continue to be sequenced,
automatic and accurate identification of genome rearrangements becomes
increasingly important, especially as high levels of rearrangement have been
observed among both eukaryotes and
prokaryotes~\citep{mammal_breakpoints,single_gene_inversions,grimm}.



In genome alignment the scope of
comparative inference expands to include questions such as, ``Which regions of
the genome are orthologous?,'' ``Which are paralogous?,'' and ``Which regions
have been acquired through horizontal gene transfer?.''  The nature of the
alignment problem varies with each question.  Shuffle-LAGAN was the first genome
aligner capable of identifying and aligning rearrangements among two genomes.
In addition to aligning single-copy regions, Shuffle-LAGAN attempts to
identify and align repetitive regions of sequence.  Our genome alignment method
represents a first step toward multiple genome
comparison in the presence of large-scale evolutionary events.  It is
capable of aligning conserved regions in the presence of genome
rearrangement, and appears to scale efficiently to long genomes. 
Further, Mauve
aligns genomes identically irrespective of their input order by identifying
multi-MUMs in subsets of the genomes and calculating a guide tree for
progressive alignment.
The remaining unaligned regions are often either repetitive, or lineage-specific
regions acquired through horizontal transfer or other means.  Repeat analysis using tools such as
RepeatMasker,  REPuter~\citep{reputer}, and FORRepeats~\citep{forrepeats} can
help to further classify unaligned regions.

%- We have presented the first system for...
%  - aligning multiple genomes in the presence of genome rearrangement
%  - method scales efficiently to long genomes
%  - independent of input genome order, guide tree calculated, clustalw blah.
%- The alignments generated by Mauve can be used for comparative genomics
%- The rearrangement structure can be used as input to programs that determine
%  rearrangement history using various methods.[cite pevzner,larget]



Much research has been devoted to inference of rearrangement history that could
lead to observed permutations in gene
order~\citep{bourquepevzner,larget,badermoret,sankoff}. The locally collinear
blocks identified during the alignment process serve as a foundation for such
methods. LCBs can naturally be reduced to the signed permutation matrix
typically used as input by these inference tools.

The evaluation of alignment quality using simulated genome evolution has yielded
several insights that will inform researchers seeking an appropriate
alignment tool.  The comparison of Mauve to Multi-LAGAN empirically confirms
the sensitivity of the CHAOS anchoring technique and LAGAN alignment method.
Multi-LAGAN successfully aligns much more divergent genomes than Mauve and is
better suited to cross-species comparison when the genomes are collinear.

Similarly, the comparison of Mauve to Shuffle-LAGAN highlights important
differences in each alignment method.  Mauve excels at aligning closely-related
sequences that have undergone modest amounts of nucleotide substitution or
inversion, consistently achieving scores above 90\% when either rate is low
relative to the other.  Conversely, Shuffle-LAGAN does best when the inversion
rate is low and nucleotide substitutions are frequent, topping out at 77.8\%
accuracy with approximately 500,000 nucleotide substitutions and 400 inversions
among the two genomes.  As previously mentioned,
Shuffle-LAGAN's difficulty anchoring in the presence of repetitive subsequences appears to cause the
anomalous result. When conducting this comparative experiment we executed
Shuffle-LAGAN as per the instructions distributed with the software, however in
the Shuffle-LAGAN paper the authors apply RepeatMasker to the genomes prior to
alignment.
RepeatMasker is not applied to the genomes by the Shuffle-LAGAN software as distributed,
and the addition of such a step may significantly alter the accuracy
of Shuffle-LAGAN alignments.


%- what was learned from simulation studies?
%  - Mauve vs. Multi-LAGAN
%    - CHAOS anchors are more sensitive
%  - Mauve vs. Shuffle-LAGAN case study:
%    - MUMs allow accurate anchoring in the face of repetitive sequence
%   - At expense of sensitivity
%   - We executed Shuffle-LAGAN as per the instructions included with the
%     software, however:
%   - Shuffle-LAGAN authors repeatMask their sequences before alignment,
%     it is not performed in the distributed Shuffle-LAGAN
%   - performing repeat masking may significantly change the quality of
%     shuffle-lagan alignments.

The design of our genome simulation system was motivated in part by our desire
to evaluate the method's ability to align genomes similar to the 9
enterobacteria.  Of course, our model simplifies or ignores many aspects of
the actual evolutionary forces at work.  Nucleotide substitution rates vary
widely throughout the genome.  Our simulation incorporated general
rate heterogeneity using a gamma distribution, $\alpha = 1$, but did not consider observed
patterns of site-specific rate heterogeneity such as third base pair
substitutions in coding regions.  Further, our model does not reflect the
phenomena of gene duplication and subsequent loss which are known to occur
frequently in the enterobacteria.  Factors such as strand bias and
site-specific rate heterogeneity for insertion, deletion, or inversion events
that may significantly alter patterns of genome evolution are not incorporated
into the model.  Despite these shortcomings, the simple genome evolver has
allowed us to characterize the limitations of our alignment system when
presented with certain well-defined patterns of evolution.  The evaluation of
alignment quality in the presence of increasing amounts of horizontal transfer
suggests that Mauve's ability to completely align genomes declines in the
presence of large lineage-specific sequence elements.  Because our method
requires homologous sequence in all genomes to anchor the alignment, lineage
specific regions larger than the maximum permitted size for progressive
alignment (10Kbp by default) remain unaligned.  Small lineage-specific regions
do not have as great an impact on alignment quality.



%  - Inferring ability to align actual sequence data based on simulations
%    - simulations account for simple evolutionary events
%   - used mutation-site rate heterogeneity, alpha == 1
%   - does not account for site-specific rate heterogeneity, such as 3rd
%     bp substitutions in genes.
%   - only minimally reflects gene duplication through repeated sampling
%     of the same donor region during H.T. insertion events
%   - does not incorporate factors such as mutation strand bias, or site-
%     specific rate heterogeneity for insertion, deletion, or inversion events.
%   - Nonetheless, simulation provides further insight into and an empirical
%     quantification of the limits of Mauve's ability to align lineage-specific
%     regions

Our experience with Mauve clearly indicates that many challenges remain in
genome alignment.  A sensitive anchoring technique that recognizes and ignores
repetitive subsequences would permit our method to be applied to more distantly
related organisms.  A method for determining breakpoints with anchors existing
in a subset of the genomes would facilitate anchored alignment of the large
lineage-specific regions currently missed.  A more comprehensive method to
assess the likelihood that a particular subset of collinear anchors represents
a true genome rearrangement could improve the sensitivity of the aligner to
small genome rearrangements.  Currently Mauve uses the somewhat ad-hoc notion
of weight to judge the likelihood of rearrangement.  Weight implicitly balances
sequence identity and size of the rearrangement:  longer rearrangements with
low sequence identity contain sufficient anchors to meet the weight cutoff,
while shorter rearrangements with higher identity will also make the cutoff
because the existence of exactly matching anchor sequences becomes more likely.
Some organisms are known to have small, local sequence rearrangements such as
reordering of protein domains in coding regions.  In such cases, the proximity
of the rearrangement to neighboring homologous sequence should clearly be
considered.   Other types of rearrangement do not exhibit locality bias:
symmetric inversions about the origin and terminus of replication and
rearrangements mediated by mobile elements are common in prokaryotes and can
move sequence to distant parts of the genome. While Shuffle-LAGAN's scoring
metric accounts for locality it is clear that not all recombination mechanisms
are subject to such a constraint. A more sophisticated scoring method may
attempt to infer the recombination mechanism suggested by a particular pattern
of anchors and then score the rearrangement based on parameters tuned to that
mechanism of recombination.

The availability and analysis of genome sequences has revealed the importance
of large-scale evolutionary events.  In light of these large-scale events, the
genome comparison problem fundamentally differs from the traditional sequence
alignment task.  By considering such large-scale events, the methods presented
here represent a significant advance toward the goal of automatic multiple genome
comparison.

%- Mauve is just an initial step
%  - Many problems in genome alignment still remain
%  - alignment of lineage-specific regions must be tackled
%  - more sensitive anchoring methods that still account for uniqueness
%  - Dealing with subset rearrangements
%  - measures for scoring the likelihood of rearrangements
%    - Mauve uses weight alone, implicitly a hybrid sequence identity/size measure
%   - difficult to draw a cutoff point in hmr comparison using weight metric
%   - small, local rearrangements are known to exist in some organisms
%   - score using a combination of some distance measure, size and sequence
%   identity

\section*{ ACKNOWLEDGMENTS }
We would like to thank Mark Craven for insightful comments and suggestions.
Funding for this research was provided by NIH Grant GM62994-02.
In addition, A. Darling was supported in part by NLM Training Grant
1T15LM007359-01.
\\
\\
\\
\\
\\
\\
\\
\\

%\section*{ FIGURES }

\begin{figure}
\centering
\epsfig{file=figures/coalescent_reprise2.eps,width=4in}
\caption{\label{fig:coalescent}
A pictorial representation of greedy breakpoint elimination in 3 genomes.
\textbf{A})  The algorithm
begins with the initial set of matching regions (multi-MUMs) represented as
connected blocks. Blocks below a genome's center line are inverted relative to the reference
sequence.
\textbf{B}) the matches are partitioned into a minimum set of
collinear blocks.  Each sequence of identically-colored blocks represents a
collinear set of matching regions.  One connecting line is drawn per collinear block.
Block 3 (yellow) has a low weight relative to other collinear blocks.
\textbf{C}) As low weight collinear blocks are removed, adjacent collinear
blocks coalesce into a single block, potentially eliminating one or more
breakpoints.  Gray regions within collinear blocks are targeted by recursive anchoring.
}
\end{figure}

\begin{figure}
\centering
\epsfig{file=figures/9_coli_unrooted_fullnames.eps,width=4.4in}
%\fig{figures/slowDisk.fig}
\bibfont
\caption{\label{fig:entero_tree} An unrooted phylogenetic tree
relating the 9 enterobacterial genomes in Table~\ref{table:9_entero}.  The tree
is a phylogenetic guide tree calculated using Neighbor-Joining by the Mauve alignment system.}
\end{figure}

\begin{figure}
\centering
\epsfig{file=figures/gradient_legend.eps,width=.75in}
\epsfig{file=figures/ntsub_indel_mauve.eps,width=2in}
\epsfig{file=figures/ntsub_indel_mlagan.eps,width=2in}
\caption{\label{fig:ntsub_indel_mauve}The performance of Mauve(left) and Multi-LAGAN(right) when aligning
sequences evolved with increasing amounts of nucleotide substitution and indels.
The multi-MUM anchoring technique employed by Mauve limits its ability to align
distantly related sequences.  Multi-LAGAN version 1.2 did not complete
the alignments of genomes without indels, resulting in the black row at the bottom.  The
substitution and indel rate observed in the enterobacteria is denoted by an asterisk(*).
}
\end{figure}


\begin{figure}
\centering
\epsfig{file=figures/gradient_legend.eps,width=.75in}
\epsfig{file=figures/ntsub_inv_mauve.eps,width=2in}
\epsfig{file=figures/ntsub_inv_slagan.eps,width=2in}
\caption{\label{fig:ntsub_inv_mauve}The performance of Mauve(left) and Shuffle-LAGAN(right) when aligning
two sequences evolved with increasing amounts of nucleotide substitution and
inversions.  Mauve is clearly more accurate than Shuffle-LAGAN at lower
substitution rates.  Shuffle-LAGAN version 1.2 did not complete some alignments
without rearrangements, resulting in black entries.  The observed substitution
and inversion rate in the enterobacteria is denoted by an asterisk(*).
}
\end{figure}

\begin{figure}
\epsfig{file=figures/gradient_legend.eps,width=.75in}
\centering \epsfig{file=figures/entero_ht.eps,width=3.4in}
%\fig{figures/slowDisk.fig}
\caption{\label{fig:entero_ht}The performance of Mauve when aligning sequences
evolved with rates similar to those observed among the group of 9 enterobacteria. 
In this experiment, the substitution, indel, and inversion frequencies were held 
constant at rates similar to those observed in the enterobacteria.  
The asterisk denotes the combination of large and
small horizontal transfer rates observed in the enterobacteria.  As
the rate of large horizontal transfer increases the amount of lineage-specific
sequence relative to backbone grows. Because Mauve can not align large lineage-
specific regions the alignment score drops.  When scored only on regions
considered backbone sequence the accuracy is consistently above 98\%. }
\end{figure}

\begin{figure}
\centering
\epsfig{file=figures/9coli_named.eps,width=5in}
\caption{\label{fig:entero_alignment}Locally collinear blocks identified
among the 9 enterobacterial genomes listed in Table~\ref{table:9_entero}.  Each
contiguously colored region is a locally collinear block, a region without
rearrangement of homologous backbone sequence.  LCBs below a genome's
center line are in the reverse complement orientation relative to the reference
genome.  Lines between genomes trace each orthologous LCB through every genome.
Large gray regions within an LCB signify the presence of lineage-specific
sequence at that site.  Each of
the 45 blocks have a minimum weight of 69.  The \textit{Shigella} and
\textit{Salmonella} genomes have undergone more genome rearrangements than the
\textit{E. coli}, likely due to the presence of specific mobile genetic elements.
The computation consumed approximately 3 hours on an 2.4GHz workstation with
1GB memory.  Figure generated by the Mauve rearrangement viewer.
}
\end{figure}

\begin{figure}
\centering
\epsfig{file=figures/hmr_jun_trans_90_chrbars.eps,width=5in}
\caption{\label{fig:hmr}Mauve visualization of locally collinear blocks identified
between concatenated chromosomes of the mouse, rat, and human genomes.  Each of
the 1,251 blocks have a minimum weight of 90.  Red vertical bars demarcate
interchromosomal boundaries.  The Mauve rearrangement viewer enables users to
interactively zoom in on regions of interest and examine the local rearrangement
structure. The computation consumed
approximately 12 hours on a 1.6GHz workstation with 2.5GB memory.
}
\end{figure}

%\section*{ TABLES }

\begin{table}
\begin{tabular}{|c|c|c|l|} \hline
Species & Genome Size & Reference \\ \hline \hline
\bibitfont{E. coli} K12 MG1655 & 4,639,221 & \citep{k12} \\ \hline
\bibitfont{E. coli} O157:H7 EDL933 & 5,524,971 & \citep{o157} \\ \hline
\bibitfont{E. coli} O157:H7 VT-2 Sakai & 5,498,450 & \citep{O157sakai} \\ \hline
\bibitfont{E. coli} CFT073 & 5,231,428 & \citep{cft073} \\ \hline
\bibitfont{S. flexneri} 2A 2457T & 4,599,354 & \citep{shig2457t} \\ \hline
\bibitfont{S. flexneri} 2A & 4,607,203 & \citep{shig2a} \\ \hline
\bibitfont{S. enterica} Typhimurium LT2 & 4,857,432 & \citep{typhimurium} \\ \hline
\bibitfont{S. enterica} Typhi CT18 & 4,809,037 & \citep{styphi} \\ \hline
\bibitfont{S. enterica} Typhi Ty2 & 4,791,961 & \citep{styphi2} \\ \hline
\hline\end{tabular}
\caption{\label{table:9_entero}
The published genome sequences of these 9 enterobacteria are a target for the
alignment system presented here.  Numerous large-scale evolutionary events such
as horizontal transfer and rearrangement are scattered throughout their genomes.
}
\end{table}

\begin{table}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|c|c|l|} \hline
&& \textbf{1} & \textbf{2} & \textbf{3} & \textbf{4} & \textbf{5} & \textbf{6} &
\textbf{7} & \textbf{8} & \textbf{9} \\ \hline
\textbf{1} & \textit{E. coli} K12 MG1655 & 1.000 & - & - & - & - & - & - & - & - \\ 
\hline
\textbf{2} & \textit{E. coli} EDL933 & 0.977 & 1.000 & - & - & - & - & - & - & - \\ 
\hline
\textbf{3} & \textit{E. coli} VT-2 Sakai & 0.978 & 1.000 & 1.000 & - & - & - & - & - & - \\ 
\hline
\textbf{4} & \textit{E. coli} CFT073 & 0.965 & 0.966 & 0.967 & 1.000 & - & - & - & - & - \\ 
\hline
\textbf{5} & \textit{S. flexneri} 2a & 0.976 & 0.975 & 0.975 & 0.963 & 1.000 & - & - & - & - \\ 
\hline
\textbf{6} & \textit{S. flexneri} 2a 2457T & 0.976 & 0.975 & 0.975 & 0.962 & 0.999 & 1.000 & - & - & - \\ 
\hline
\textbf{7} & \textit{S.} Typhimurium & 0.794 & 0.793 & 0.793 & 0.793 & 0.791 & 0.791 & 1.000 & - & - \\ 
\hline
\textbf{8} & \textit{S.} typhi CT18 & 0.792 & 0.791 & 0.791 & 0.792 & 0.790 & 0.789 & 0.981 & 1.000 & - \\ 
\hline
\textbf{9} & \textit{S.} typhi Ty2 & 0.793 & 0.793 & 0.793 & 0.793 & 0.791 & 0.791 & 0.984 & 0.996 & 1.000 \\ 
\hline
\hline\end{tabular}
\caption{\label{table:entero_identity} Identity matrix 
for 2.86Mbp shared backbone regions among the 9 enterobacteria listed in 
Table~\ref{table:9_entero}.  Although an average of only 58\% of the genomes are
conserved across species, the level of sequence identity is remarkably high,
suggesting that horizontal transfer and differential gene loss may account for
the majority of phenotypic diversity among bacteria in this group.}  
\end{table}

\section*{}
\fontfamily{ptm}\selectfont
\bibliographystyle{apalike}
%\bibliographystyle{harvard}
%\bibliographystyle{plainnat}
\bibliography{genres}

\section*{ WEBSITE REFERENCES }

\end{document}
